### ISSUE TITLE
Coal/Charcoal blocks smelt way less items than expected, in an electric furnace at least

### ISSUE THREAD

**Issue by heyjudek** [TechReborn/TechReborn#2403]
In short, while a piece of coal/charcoal would smelt 20 items, it should be expected that its block variant would smelt 200, but it ends up being in 70s which is even worse than what one would expect in a regular furnace.
The problem is that I could not locate a problem in the math. A piece of coal/charcoal provides 4k energy and smelting one item takes 0.2k. Therefore, 4/0.2=20
A coal/charcoal block fill the generator to capacity (10k) and it still shows 75% left in the reserve, which adds up to a total of 40k, but it unfortunately falls far short of the expected 200 smelted items.

**Comment by Technici4n** [comment]
Maybe you should connect the generator to a bigger storage to measure how much energy is produced exactly? (MFE maybe?)

**Comment by heyjudek** [comment]
Well, the question is, why is this even necessary? It recharges the generator fully and has 75% left, indicating it has 40k energy. But during the smelting process it falls far short of smelting 10x the items a piece of coal/charcoal would smelt.
I am sure you can reproduce this issue too. Place an electric furnace with a generator next to it, they should both be uncharged. Then add a piece of coal/charcoal block and see how many items you will manage to smelt, it will probably be around 70.

**Comment by drcrazy** [comment]
Solid fuel generator produces 10 E\t according to default config.
Electric furnace uses 1E\t for smelting items.

In theory there could be situation when not whole 10 E\t added per tick
https://github.com/TechReborn/TechReborn/blob/44a0313361a3feb09360bc6acf30e98c8f6be4d7/src/main/java/techreborn/blockentity/generator/basic/SolidFuelGeneratorBlockEntity.java#L87

Like you have 9999 out of 10k. And that will decrease burn time by one tick but will add only 1 energy during that tick and not 10

**Comment by heyjudek** [comment]
In that case, what is the cause of such a low smelting rate with coal/charcoal blocks? It just doesn't make sense if a coal smelts 20 but a coal block smelts 70 while it should be 200

**Comment by Technici4n** [comment]
The energy is voided if the fuel is burning but the buffer is full... and 1 E/t is enough to keep draining the buffer, so that means 10 - 1 = 9 EU/t will get voided.

**Comment by heyjudek** [comment]
Then I guess it is not recommended to use coal/charcoal blocks with generators?
Is this something that would be fixed in the future?

**Comment by Technici4n** [comment]
There is nothing to fix? Use a bigger buffer, or use more energy.

**Comment by drcrazy** [comment]
Well, we can try to add energy only when internal buffer can accept all 10 Energy

But I expect another issue "Coal generator stuck at 9.999 out of 10.000"

**Comment by heyjudek** [comment]
Well, I can't really speak about the modding side of it, but I think it should be nice as an end result if coal/charcoal blocks behave as expected, as in smelting 10x the items coal/charcoal pieces would.

**Comment by heyjudek** [comment]
> There is nothing to fix? Use a bigger buffer, or use more energy.

I am confused, block of coal/charcoal already fills a generator to maximum capacity + 75% leftover. The math seems to be broken here. The problem is that when that "leftover" energy is being transferred to the generator charge, it is being transferred at a much worse rate than expected. If I added even more energy, wouldn't that make this problem even bigger?

**Comment by drcrazy** [comment]
```
 The problem is that when that "leftover" energy is being transferred to the generator charge, it is being transferred at a much worse rate than expected.
 ```
 
This is not transfer issue. Put 10 smelteries to one coal generator and you will be fine. Or put buffer big enough to hold energy of one block of coal..

**Comment by heyjudek** [comment]
> ```
>  The problem is that when that "leftover" energy is being transferred to the generator charge, it is being transferred at a much worse rate than expected.
> ```
> 
> This is not transfer issue. Put 10 smelteries to one coal generator and you will be fine. Or put buffer big enough to hold energy of one block of coal..

So, if I had I generator that could cold 40k energy without showing 75% as extra, it would smelt 200 items?
I am still confused as to how this is not a transfer issue. As an end result, a fuel providing 40k energy smelts roughly a third of the items it is supposed to smelt. And when 1 block of coal/charcoal is added to a generator it doesn't show that it can't hold the amount of energy a block of coal/charcoal is trying to add. It is 10k + 75%.
The reason I suggested this might be a transfer issue is because that 75% is 30k energy, but it ends up being roughly 10k when it comes to its smelting capacity.

**Comment by Technici4n** [comment]
Notice how a vanilla furnace will burn an entire piece of coal even if you only put one cobblestone in the furnace, and the rest is "voided". Exactly the same happens here if the generator consumes the piece of coal to start producing energy, but has nowhere to send it. Just connect enough machines to use the power, or use a bigger battery. Do not let the generator fill up or the excess energy will be voided.

**Comment by heyjudek** [comment]
> Notice how a vanilla furnace will burn an entire piece of coal even if you only put one cobblestone in the furnace
But if you do put 8 items to smelt, one piece of coal will smelt them all, right? So, you have a "choice" to let that piece of coal/charcoal go to waste unless you put 8 items in total. But I do not even have this choice to prevent this waste even if I provide 200 items to smelt.
If I put 200 items with a hopper setup, 1 piece of coal/charcoal will not even smelt half of them.


> Exactly the same happens here if the generator consumes the piece of coal to start producing energy, but has nowhere to send it. 

This is what makes me confused and that's why I reported this as a bug. 1 coal/charcoal block has 40k while a generator has a capacity of 10k. So, 1 generator can only accomodate 25% of a coal/charcoal block's smelting capacity. Are we on the same page so far? 
But, you know the fire icon that has a percentage indicator? It does show 75% "leftover" after a generator is recharged from 0 to 10k. 
Now, my question is, as long as normal math is involved, if recharging from 0 to 10k takes 25%. Doesn't that mean that 75% would equal 30k energy which means 150 items smelted? (30000/200)

So in short,
10k energy = 25% smelts 50 items
30k energy "leftover" = 75% but smelts 20-25 items? How is this not a bug?
In total, 1 coal/charcoal block falls short of smelting 75 items.

**Comment by Wdavery** [comment]
There is clearly some miscommunication here. The issue here, @heyjudek, is that your setup does not have the energy storage capacity for an entire block of coal. 
In vanilla, one block of coal to smelt one cobblestone>stone will result in an incredible waste of fuel.
The same is true here. Energy is being wasted because it has nowhere to go.

To fix this issue for yourself, simply add a battery box between the generator and the electric furnace. This will allow the entire 40kE to be stored and should result in the full 200 items being smelted.

This is ultimately not a bug in the mod, but a flaw in the design of your system. (like putting a block of coal in a furnace to smelt a handful of items, would be a flaw in a vanilla system)

Try out what I suggested and report back as I think you'll find it's safe to close this issue. But this may be something that requires some information in the wiki to help other players with the same issue, so thank you for bringing it up.

**Comment by drcrazy** [comment]
Better buffer should solve issue.
